File: v5-go-implementation/internal/feedback/feedback.go
// Package feedback implements an append-only JSONL routing record store
// with per-model analytics.
package feedback

import (
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// RoutingRecord is one persisted routing call.
type RoutingRecord struct {
	Timestamp      string  `json:"timestamp"`
	TaskType       string  `json:"task_type"`
	Complexity     float64 `json:"complexity"`
	Tier           int     `json:"tier"`
	ModelUsed      string  `json:"model_used"`
	LatencyMs      int     `json:"latency_ms"`
	CostUnits      float64 `json:"cost_units"`
	QualityEval    float64 `json:"quality_eval"`
	PromptLen      int     `json:"prompt_len"`
	FailoverUsed   bool    `json:"failover_used"`
	FailoverFrom   string  `json:"failover_from"`
	FailoverReason string  `json:"failover_reason"`
}

// Store is a thread-safe append-only JSONL record store.
type Store struct {
	path    string
	records []RoutingRecord
	mu      sync.Mutex
}

// New opens (or creates) the JSONL file at path.
func New(path string) (*Store, error) {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return nil, err
	}
	s := &Store{path: path}
	if err := s.load(); err != nil {
		return nil, err
	}
	return s, nil
}

func (s *Store) load() error {
	f, err := os.Open(s.path)
	if os.IsNotExist(err) {
		return nil
	}
	if err != nil {
		return err
	}
	defer f.Close()
	dec := json.NewDecoder(f)
	for dec.More() {
		var r RoutingRecord
		if err := dec.Decode(&r); err == nil {
			s.records = append(s.records, r)
		}
	}
	return nil
}

// Record appends a routing record to the JSONL file.
func (s *Store) Record(
	taskType string,
	complexity float64,
	tier int,
	modelUsed string,
	latencyMs int,
	outputTokens int,
	costPer1k float64,
	promptLen int,
	failoverUsed bool,
	failoverFrom string,
	failoverReason string,
) {
	qualityEval := HeuristicQuality(latencyMs, outputTokens)
	costUnits := costPer1k * float64(outputTokens) / 1000.0
	rec := RoutingRecord{
		Timestamp:      time.Now().UTC().Format(time.RFC3339Nano),
		TaskType:       taskType,
		Complexity:     math.Round(complexity*1000) / 1000,
		Tier:           tier,
		ModelUsed:      modelUsed,
		LatencyMs:      latencyMs,
		CostUnits:      math.Round(costUnits*1e6) / 1e6,
		QualityEval:    math.Round(qualityEval*1000) / 1000,
		PromptLen:      promptLen,
		FailoverUsed:   failoverUsed,
		FailoverFrom:   failoverFrom,
		FailoverReason: failoverReason,
	}
	s.mu.Lock()
	s.records = append(s.records, rec)
	s.mu.Unlock()

	// Append to file (best-effort — don't block callers on I/O error)
	if f, err := os.OpenFile(s.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644); err == nil {
		enc := json.NewEncoder(f)
		_ = enc.Encode(rec)
		f.Close()
	}
}

// HeuristicQuality computes a quality proxy from latency and token count.
func HeuristicQuality(latencyMs int, outputTokens int) float64 {
	if outputTokens == 0 {
		return 0.0
	}
	latSec := math.Max(1, float64(latencyMs)/1000)
	tps := float64(outputTokens) / latSec
	q := 0.5 + tps/60
	if q > 1.0 {
		return 1.0
	}
	return q
}

// Summary returns aggregated analytics over all stored records.
func (s *Store) Summary() map[string]interface{} {
	s.mu.Lock()
	recs := make([]RoutingRecord, len(s.records))
	copy(recs, s.records)
	s.mu.Unlock()

	if len(recs) == 0 {
		return map[string]interface{}{"total_calls": 0}
	}

	var totalCost, totalQuality float64
	var totalLatency int
	failoverCount := 0
	byTier := map[int]int{}
	type modelAcc struct {
		calls, totalLatency int
		totalCost, totalQuality float64
	}
	byModel := map[string]*modelAcc{}

	for _, r := range recs {
		totalCost += r.CostUnits
		totalQuality += r.QualityEval
		totalLatency += r.LatencyMs
		if r.FailoverUsed {
			failoverCount++
		}
		byTier[r.Tier]++
		if byModel[r.ModelUsed] == nil {
			byModel[r.ModelUsed] = &modelAcc{}
		}
		acc := byModel[r.ModelUsed]
		acc.calls++
		acc.totalLatency += r.LatencyMs
		acc.totalCost += r.CostUnits
		acc.totalQuality += r.QualityEval
	}
	n := len(recs)

	modelStats := map[string]interface{}{}
	for label, acc := range byModel {
		c := acc.calls
		modelStats[label] = map[string]interface{}{
			"calls":       c,
			"avg_latency": acc.totalLatency / c,
			"avg_cost":    math.Round(acc.totalCost/float64(c)*1e6) / 1e6,
			"avg_quality": math.Round(acc.totalQuality/float64(c)*1000) / 1000,
		}
	}

	return map[string]interface{}{
		"total_calls":    n,
		"total_cost":     math.Round(totalCost*1e6) / 1e6,
		"avg_quality":    math.Round(totalQuality/float64(n)*1000) / 1000,
		"avg_latency_ms": totalLatency / n,
		"failover_count": failoverCount,
		"calls_by_tier":  byTier,
		"by_model":       modelStats,
	}
}

File: routing-v4/src/feedback.py
"""
feedback.py  —  routing-v4
Append-only JSONL store for routing records.
Enhanced with per-model statistics, failover tracking, and threshold history.
"""

from __future__ import annotations
import json
import os
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from pathlib import Path
from collections import defaultdict


@dataclass
class RoutingRecord:
    timestamp:      str
    task_type:      str
    complexity:     float
    tier:           int
    model_used:     str
    latency_ms:     int
    cost_units:     float
    quality_eval:   float
    prompt_len:     int
    failover_used:  bool   = False
    failover_from:  str    = ""
    failover_reason: str   = ""


class FeedbackStore:
    """Append-only JSONL store with richer per-model analytics."""

    def __init__(self, path: str = "output/feedback.jsonl") -> None:
        self._path = Path(path)
        self._path.parent.mkdir(parents=True, exist_ok=True)
        self._records: list[RoutingRecord] = self._load()

    def _load(self) -> list[RoutingRecord]:
        if not self._path.exists():
            return []
        records = []
        with self._path.open() as fh:
            for line in fh:
                line = line.strip()
                if line:
                    try:
                        d = json.loads(line)
                        # Backfill missing fields from older records
                        d.setdefault("failover_used",   False)
                        d.setdefault("failover_from",   "")
                        d.setdefault("failover_reason", "")
                        records.append(RoutingRecord(**d))
                    except Exception:
                        pass
        return records

    def record(
        self,
        task_type:       str,
        complexity:      float,
        tier:            int,
        model_used:      str,
        latency_ms:      int,
        output_tokens:   int,
        cost_per_1k:     float,
        prompt_len:      int,
        failover_used:   bool = False,
        failover_from:   str  = "",
        failover_reason: str  = "",
    ) -> None:
        quality_eval = self._heuristic_quality(latency_ms, output_tokens)
        cost_units   = cost_per_1k * output_tokens / 1000.0
        rec = RoutingRecord(
            timestamp=datetime.now(timezone.utc).isoformat(),
            task_type=task_type,
            complexity=round(complexity, 3),
            tier=tier,
            model_used=model_used,
            latency_ms=latency_ms,
            cost_units=round(cost_units, 6),
            quality_eval=round(quality_eval, 3),
            prompt_len=prompt_len,
            failover_used=failover_used,
            failover_from=failover_from,
            failover_reason=failover_reason,
        )
        self._records.append(rec)
        with self._path.open("a") as fh:
            fh.write(json.dumps(asdict(rec)) + "\n")

    @staticmethod
    def _heuristic_quality(latency_ms: int, output_tokens: int) -> float:
        if output_tokens == 0:
            return 0.0
        tps = output_tokens / max(1, latency_ms / 1000)
        return min(1.0, 0.5 + tps / 60)

    def summary(self) -> dict:
        if not self._records:
            return {"total_calls": 0}
        total_cost     = sum(r.cost_units   for r in self._records)
        avg_quality    = sum(r.quality_eval for r in self._records) / len(self._records)
        avg_latency    = sum(r.latency_ms   for r in self._records) / len(self._records)
        failover_count = sum(1 for r in self._records if r.failover_used)

        by_tier: dict[int, int] = {}
        for r in self._records:
            by_tier[r.tier] = by_tier.get(r.tier, 0) + 1

        by_model: dict[str, dict] = defaultdict(
            lambda: {"calls": 0, "total_latency_ms": 0, "total_cost": 0.0, "total_quality": 0.0}
        )
        for r in self._records:
            s = by_model[r.model_used]
            s["calls"]            += 1
            s["total_latency_ms"] += r.latency_ms
            s["total_cost"]       += r.cost_units
            s["total_quality"]    += r.quality_eval

        model_stats = {}
        for label, s in by_model.items():
            c = s["calls"]
            model_stats[label] = {
                "calls":       c,
                "avg_latency": round(s["total_latency_ms"] / c),
                "avg_cost":    round(s["total_cost"] / c, 6),
                "avg_quality": round(s["total_quality"] / c, 3),
            }

        return {
            "total_calls":    len(self._records),
            "total_cost":     round(total_cost, 6),
            "avg_quality":    round(avg_quality, 3),
            "avg_latency_ms": round(avg_latency),
            "failover_count": failover_count,
            "calls_by_tier":  by_tier,
            "by_model":       model_stats,
        }
